components/bluetooth/lib.rs

```diff
@@ -678,328 +648,151 @@ impl BluetoothManager {
-    fn get_primary_service(&mut self,
-                           device_id: String,
-                           uuid: String,
-                           sender: IpcSender<BluetoothResponseResult>) {
-        // Step 5.
-        if !self.cached_devices.contains_key(&device_id) {
-            return drop(sender.send(Err(BluetoothError::InvalidState)));
-        }
+    fn get_gatt_children(&mut self,
+                         id: String,
+                         uuid: Option<String>,
+                         single: bool,
+                         child_type: GATTType,
+                         sender: IpcSender<BluetoothResponseResult>) {
         let mut adapter = get_adapter_or_return_error!(self, sender);
-        if !self.allowed_services.get(&device_id).map_or(false, |s| s.contains(&uuid)) {
-            return drop(sender.send(Err(BluetoothError::Security)));
-        }
-        let services = self.get_gatt_services_by_uuid(&mut adapter, &device_id, &uuid);
-
-        // Step 6.
-        for service in services {
-            if service.is_primary().unwrap_or(false) {
-                if let Ok(uuid) = service.get_uuid() {
-                    return drop(sender.send(
-                        Ok(BluetoothResponse::GetPrimaryService(
-                            BluetoothServiceMsg {
-                                uuid: uuid,
-                                is_primary: true,
-                                instance_id: service.get_id(),
-                            }
-                        ))
-                    ));
+        match child_type {
+            GATTType::PrimaryService => {
+                // Step 5.
+                if !self.cached_devices.contains_key(&id) || !self.address_to_id.values().any(|v| v == &id) {
+                    return drop(sender.send(Err(BluetoothError::InvalidState)));
                 }
-            }
-        }
-        // Step 7.
-        return drop(sender.send(Err(BluetoothError::NotFound)));
-    }
-
-    // https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothremotegattserver-getprimaryservices
-    // https://webbluetoothcg.github.io/web-bluetooth/#getgattchildren
-    fn get_primary_services(&mut self,
-                            device_id: String,
-                            uuid: Option<String>,
-                            sender: IpcSender<BluetoothResponseResult>) {
-        // Step 5.
-        if !self.cached_devices.contains_key(&device_id) {
-            return drop(sender.send(Err(BluetoothError::InvalidState)));
-        }
-        let mut adapter = get_adapter_or_return_error!(self, sender);
-        let services = match uuid {
-            Some(ref id) => {
-                if !self.allowed_services.get(&device_id).map_or(false, |s| s.contains(id)) {
-                    return drop(sender.send(Err(BluetoothError::Security)))
+                // Step 6.
+                if let Some(uuid) = uuid.clone() {
```

Comment:
Maybe try with `if let Some(ref uuid) = uuid`, I think this way the `clone()` can be omitted.